### ISSUE TITLE
RouterProvider not working in tests (Vitest)

### ISSUE THREAD

**Issue by rothsandro** [remix-run/react-router#12512]
### I'm using React Router as a...

library

### Reproduction

1. Clone the repo https://github.com/rothsandro/repro.react-router-provider-vitest
2. Install dependencies `npm install`
3. Run the tests `npm run test`

There is a simple test that renders the `RouterProvider` with a memory router. 
We then click on a button that triggers a navigation with `flushSync: true` using the `useNavigate()` hook.

### System Info

```shell
System:
    OS: macOS 14.2.1
    CPU: (10) arm64 Apple M1 Max
    Memory: 35.43 GB / 64.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.17.0 - ~/.nvm/versions/node/v20.17.0/bin/node
    npm: 10.8.2 - ~/.nvm/versions/node/v20.17.0/bin/npm
    pnpm: 9.12.3 - /opt/homebrew/bin/pnpm
    bun: 1.0.36 - ~/.bun/bin/bun
  Browsers:
    Brave Browser: 131.1.73.97
    Chrome: 130.0.6723.70
    Safari: 17.2.1
  npmPackages:
    react-router: ^7.0.2 => 7.0.2 
    vite: ^5.4.11 => 5.4.11
```

### Used Package Manager

npm

### Expected Behavior

The test should run successfully and without error or warning.

### Actual Behavior

The tests throws an error:

> useNavigate() may be used only in the context of a &lt;Router&gt; component.

The upgrade guide mentions that for non-DOM the router provider should be imported from `react-router` instead of `react-router/dom`. The repro uses Vitest with happy-dom so it's a DOM environment. There is a 2nd test that uses the router provider from `react-router`. This solves the error from above but will log a warning to the console that the router provider from `react-router/dom` should be used for `flushSync` (skip the failing test to see the warning).

**Comment by steven-muniz** [comment]
I am literally having this same issue! Weird part is all the test work fine in local, once I load them to the CI/CD that's when I see if fail.

**Comment by hermaneyd** [comment]
In a similar case, adding `react-router-dom` dependency and importing everything from that package solved the error.

_package.json_
```json
{
...
  "dependencies": {
    ...
    "react-router": "7.1.1",
    "react-router-dom": "7.1.1",
  }
}
```

_...test.tsx_
```javascript
import { createMemoryRouter, RouterProvider, useNavigate } from "react-router-dom";
```

**Comment by rothsandro** [comment]
@hermaneyd That's the old package and is only used for a smooth upgrade.

**Comment by samuelsantia-vass** [comment]
Hey, I have the same issue upgrading one of my projects. I've just fixed it updating my imports to use in all cases `react-router` instead `react-router-dom`

It might be something on how vite treat the modules, but yes is very weird 🤔 

I'm using the `createRoutesStub` helper

**Comment by rothsandro** [comment]
@brophdawg11 I just installed `react-router@0.0.0-experimental-b2bc694a6` in my reproduction and the test is green now 🤔

**Comment by brophdawg11** [comment]
oh - weird.  Let me try again in my clone - maybe I messed up my install.  I was doing a lot of local copying around of built files and such but I thought I did a fresh npm install of the experimental at the end.

**Comment by brophdawg11** [comment]
Hm - yeah I'm still getting the error.  If you pull my branch fresh and `npm install && npm run test` does it work for you?

[git@github.com:brophdawg11/repro.react-router-provider-vitest.git](https://github.com/brophdawg11/repro.react-router-provider-vitest/tree/rr-experimental)

**Comment by rothsandro** [comment]
I don't get the original context anymore with your branch.

I get an error from testing library but that's just because of the missing clean up, adding `afterEach(() => cleanup());` solves the problem.

The warning `You provided the flushSync option...` from the 2nd test stays but that's expected.

**Comment by brophdawg11** [comment]
That's super odd - I still get it on a fresh clone of my fork branch.  Going to check in with a few other team members and see what behavior they get and if we can figure out why it would still be failing for me

**Comment by rothsandro** [comment]
@brookslybrand That's the testing library error because of the missing cleanup (I mentioned above), so the fix actually works. Just skip the 2nd test and you'll see the first test passes.

**Comment by rothsandro** [comment]
@brookslybrand Hm, weird. Could a different node/npm version be the problem?

**Comment by rothsandro** [comment]
Looks like it depends on the node version, just tested it with different versions.
The fix works with Node 21.5.0 and also with 22.5.1 but the error still occurs with 22.15.0

**Comment by rothsandro** [comment]
@markdalgleish Hm not sure. The other issue mentions two conditions which don't apply here:

> - if your project depends on react-router and a dependency which in turn has a peer-dependency on react router
> - It also only affects node 22.12 and newer

**Comment by brophdawg11** [comment]
@markdalgleish I do think this is related to #12785.  With the fix in https://github.com/remix-run/react-router/pull/13497, the test no longer fails on node 22.11 and continues to fail on 22.12.  I just confirmed against a fresh experimental release on my form branch: https://github.com/brophdawg11/repro.react-router-provider-vitest/tree/rr-experimental

You can clone that and:

```sh
> n 22.11 && npm ci && npm run test ## ✅
> n 22.12 && npm ci && npm run test ## 💥
```

Therefore, I'm going to merge my linked PR and close this ticket out as resolved and we can continue to track the node 22.12 issue in #12785.

**Comment by brophdawg11** [comment]
@markdalgleish To clarify - I think this bundling issue was standalone and caused breakages on node <22.12.  Once we fixed that, I think we now exposed the same issue as #12785 where node 22.12 and above seem to end up with both CJS and ESM modules being loaded.
